### PR TITLE
implement `dpnp.piecewise`

### PR DESCRIPTION
**Pull request by vtavana**
In this PR, `dpnp.piecewise` is implemented.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] Have you added documentation for your changes, if necessary?
- [x] Have you added your changes to the changelog?

### CODE REVIEW

File: dpnp/dpnp_iface_functional.py [hunk 1]
```diff
@@ -266,3 +273,145 @@ def apply_over_axes(func, a, axes):
...
+    Evaluate a piecewise-defined function.
+
+    Given a set of conditions and corresponding functions, evaluate each
+    function on the input data wherever its condition is true.
+
+    For full documentation refer to :obj:`numpy.piecewise`.
+
+    Parameters
+    ----------
+    x : : {dpnp.ndarray, usm_ndarray}
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_functional.py]:
```suggestion
    x : {dpnp.ndarray, usm_ndarray}
```

File: dpnp/dpnp_iface_functional.py [hunk 2]
```diff
@@ -266,3 +273,145 @@ def apply_over_axes(func, a, axes):
...
+
+    for condition, func in zip(condlist, funclist):
+        if callable(func):
+            raise NotImplementedError(
+                "Callable functions are not supported currently"
+            )
+        if isinstance(func, dpnp.ndarray):
+            func = func.astype(x.dtype)
+
+        # TODO: possibly can use func.item() to make sure that func is always
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_functional.py]: Would it make sense to use `dpnp.where` here instead? Or are there any limitations or performance concerns against that?

File: dpnp/dpnp_iface_functional.py [hunk 3]
```diff
@@ -266,3 +273,145 @@ def apply_over_axes(func, a, axes):
...
+    >>> x = np.linspace(-2.5, 2.5, 6)
+    >>> np.piecewise(x, [x < 0, x >= 0], [-1, 1])
+    array([-1., -1., -1.,  1.,  1.,  1.])
+
+    """
+    dpnp.check_supported_arrays_type(x)
+    if isinstance(condlist, dpnp.ndarray) and condlist.ndim in [0, 1]:
+        condlist = [condlist]
+    elif dpnp.isscalar(condlist) or (
+        dpnp.isscalar(condlist[0]) and x.ndim != 0
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_functional.py:342]: There is no check that `condlist` is exactly a list. Do we need that or it is assuming to accept any sequence?

Comment (vtavana) [reply dpnp/dpnp_iface_functional.py:342]: In practice, `condlist` can be a list, tuple or ndarray in both NumPy and CuPy. And `dpnp` follows the same practice.

```python
import numpy
a = numpy.linspace(-2.5, 2.5, 6)
numpy.piecewise(a, [a < 0, a >= 0], [-1, 1])  # condlist is list
# array([-1., -1., -1.,  1.,  1.,  1.])

numpy.piecewise(a, (a < 0, a >= 0), [-1, 1])  # condlist is tuple
# array([-1., -1., -1.,  1.,  1.,  1.])

numpy.piecewise(a, numpy.array([a < 0, a >= 0]), [-1, 1])  # condlist is array
# array([-1., -1., -1.,  1.,  1.,  1.])

import cupy
x = cupy.array(a)

cupy.piecewise(x, [x < 0, x >= 0], [-1, 1])  # condlist is list
# array([-1., -1., -1.,  1.,  1.,  1.])

cupy.piecewise(x, (x < 0, x >= 0), [-1, 1])  # condlist is tuple
# array([-1., -1., -1.,  1.,  1.,  1.])

cupy.piecewise(x, cupy.array([x < 0, x >= 0]), [-1, 1])  # condlist is array
# array([-1., -1., -1.,  1.,  1.,  1.])

import dpnp
ia = dpnp.array(a)
dpnp.piecewise(ia, [ia < 0, ia >= 0], [-1, 1])  # condlist is list
# array([-1., -1., -1.,  1.,  1.,  1.])

dpnp.piecewise(ia, (ia < 0, ia >= 0), [-1, 1])  # condlist is tuple
# array([-1., -1., -1.,  1.,  1.,  1.])

dpnp.piecewise(ia, dpnp.array([ia < 0, ia >= 0]), [-1, 1])  # condlist is array
# array([-1., -1., -1.,  1.,  1.,  1.])
```

File: dpnp/backend/extensions/functional/piecewise.cpp [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    }
+
+    if (!dpctl::utils::queues_are_compatible(
+            exec_q, {condition.get_queue(), result.get_queue()}))
+    {
+        throw py::value_error(
+            "Execution queue is not compatible with allocation queue.");
+    }
+
+    const bool is_result_c_contig = result.is_c_contiguous();
```
Comment (vlad-perevezentsev) [inline dpnp/backend/extensions/functional/piecewise.cpp]: I think `condition` should also be `C-contig` to keep the kernel running correctly

```
In [1]: import dpnp

In [2]: import numpy

In [3]: a = numpy.arange(8).reshape(2,2,2)

In [4]: cond = numpy.array(a>3, order="F")

In [5]: a_dp = dpnp.array(a)

In [6]: cond_dp = dpnp.array(cond, order="F")

In [7]: numpy.piecewise(a, [cond], [1])
Out[7]: 
array([[[0, 0],
        [0, 0]],

       [[1, 1],
        [1, 1]]])

In [8]: dpnp.piecewise(a_dp,[cond_dp],[1])
Out[8]: 
array([[[0, 1],
        [0, 1]],

       [[0, 1],
        [0, 1]]])

```

File: dpnp/backend/extensions/functional/piecewise.cpp [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    dpctl::tensor::validation::CheckWritable::throw_if_not_writable(result);
+
+    const int res_nd = result.get_ndim();
+    const int cond_nd = condition.get_ndim();
+    if (res_nd != cond_nd) {
+        throw py::value_error(
+            "Condition and result arrays must have the same dimension.");
+    }
+
+    if (!dpctl::utils::queues_are_compatible(
```
Comment (vlad-perevezentsev) [inline dpnp/backend/extensions/functional/piecewise.cpp]: Could you use `check_queue` from common validation utils?

File: dpnp/backend/extensions/functional/piecewise.cpp [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            exec_q, {condition.get_queue(), result.get_queue()}))
+    {
+        throw py::value_error(
+            "Execution queue is not compatible with allocation queue.");
+    }
+
+    const bool is_result_c_contig = result.is_c_contiguous();
+    if (!is_result_c_contig) {
```
Comment (vlad-perevezentsev) [inline dpnp/backend/extensions/functional/piecewise.cpp]: same
`check_c_contig`

File: dpnp/backend/extensions/functional/piecewise.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        throw py::value_error("The result array is not c-contiguous.");
+    }
+
```
Comment (vlad-perevezentsev) [inline dpnp/backend/extensions/functional/piecewise.cpp]: Should we check for no memory overlap?

File: dpnp/dpnp_iface_functional.py [hunk 8]
```diff
@@ -266,3 +273,145 @@ def apply_over_axes(func, a, axes):
...
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_functional.py:342]: Does it make sense to raise an error like `ValueError` if `condlist` is empty?

```
dpnp.piecewise(x, [], [1])
# output
IndexError: list index out of range
```

File: dpnp/dpnp_iface_functional.py [hunk 9]
```diff
@@ -266,3 +273,145 @@ def apply_over_axes(func, a, axes):
...
+
+    dpnp.check_supported_arrays_type(*condlist)
+    if dpnp.is_supported_array_type(funclist):
+        usm_type, exec_q = get_usm_allocations([x, *condlist, funclist])
+    else:
+        usm_type, exec_q = get_usm_allocations([x, *condlist])
+
+    condlen = len(condlist)
+    try:
+        funclen = len(funclist)
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_functional.py:373]: This check passes if we pass `funclist` as a string
```
dpnp.piecewise(x, [x>0], "q")
RuntimeError: Unable to cast Python instance of type <class 'str'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```

File: dpnp/dpnp_iface_functional.py [hunk 10]
```diff
@@ -266,3 +273,145 @@ def apply_over_axes(func, a, axes):
...
+            )
+        result = dpnp.full(
+            x.shape, func, dtype=x.dtype, usm_type=usm_type, sycl_queue=exec_q
+        )
+    else:
+        raise ValueError(
+            f"with {condlen} condition(s), either {condlen} or {condlen + 1} "
+            "functions are expected"
+        )
+
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_functional.py:397]: Please handle this case as well
```
x = dpnp.ones(3)
dpnp.piecewise(x, [x>0], dpnp.array([[1,2,3]]))
# output 
RuntimeError: Unable to cast Python instance of type <class 'dpnp.dpnp_array.dpnp_array'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```

File: dpnp/dpnp_iface_functional.py (hunk 2 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_functional.py]: Using `dpnp.where` instead of kernel is ~2x slower
```
import numpy, dpnp
a = numpy.linspace(-5, 5, 4096*4096)
%timeit numpy.piecewise(a, [a < 0, a >= 0], [-1, 1])
# 91.4 ms ± 148 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
ia = dpnp.array(a)
%timeit res = dpnp.piecewise(ia, [ia < 0, ia >= 0], [-1, 1]); res.sycl_queue.wait()
# 1.64 ms ± 4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# 3.06 ms ± 1.17 ms per loop (mean ± std. dev. of 7 runs, 100 loops each) # using dpnp.where
```

But it keeps the code simple and supports broadcasting
To improve performance, you can try to avoid `dpnp.zeros_like()` and `dpnp.full()` in `condlen == funclen` and `condlen + 1 == funclen` cases.

Instead, create one `def_value`, allocate result once `dpnp.empty_like(...)`, and update it in-place with `dpnp.where(... out=result)` in the loop using any flags to detect the first iteration
First call - `dpnp.where(condition, func, def_value, out=result)` to fill non-matching elements with `def_value`
Other calls -  `dpnp.where(condition, func, result, out=result)  to keep existing values and overwrite only where condition is True.`

Then I get on PVC for float32

```
# 5.97 ms ± 109 μs per loop (mean ± std. dev. of 7 runs, 100 loops each) # kernel
# vs
# 6.47 ms ± 8.89 μs per loop (mean ± std. dev. of 7 runs, 100 loops each) # optimized where 
```

File: dpnp/dpnp_iface_functional.py (hunk 10 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_functional.py:397]: Resolved

File: dpnp/dpnp_iface_functional.py [hunk 11]
```diff
@@ -266,3 +270,141 @@ def apply_over_axes(func, a, axes):
...
+        default_value = x_dtype.type(0)
+    elif condlen + 1 == funclen:
+        # default value is the last element of funclist
+        default_value = funclist[-1]
+        if callable(default_value):
+            raise NotImplementedError(
+                "Callable functions are not supported currently"
+            )
+        if isinstance(default_value, dpnp.ndarray):
+            default_value = default_value.astype(x_dtype)
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_functional.py]: To avoid unnecessary copying, you can set `copy=False` here

File: dpnp/dpnp_iface_functional.py [hunk 12]
```diff
@@ -266,3 +270,141 @@ def apply_over_axes(func, a, axes):
...
+            "functions are expected"
+        )
+
+    for condition, func in zip(condlist, funclist):
+        if callable(func):
+            raise NotImplementedError(
+                "Callable functions are not supported currently"
+            )
+        if isinstance(func, dpnp.ndarray):
+            func = func.astype(x_dtype)
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_functional.py]: the same `copy=False` will be better

File: dpnp/dpnp_iface_functional.py [hunk 13]
```diff
@@ -266,3 +270,141 @@ def apply_over_axes(func, a, axes):
...
+        usm_type, exec_q = get_usm_allocations([x, *condlist, funclist])
+    else:
+        usm_type, exec_q = get_usm_allocations([x, *condlist])
+
+    result = dpnp.empty_like(x, usm_type=usm_type, sycl_queue=exec_q)
+
+    condlen = len(condlist)
+    try:
+        if isinstance(funclist, str):
+            raise TypeError
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_functional.py]: No coverage

File: dpnp/tests/test_functional.py [hunk 14]
```diff
@@ -65,3 +75,244 @@ def custom_func(x, axis):
...
+        expected = numpy.piecewise(a, ([True, False], [False, True]), [1, -4])
+        result = dpnp.piecewise(ia, ([True, False], [False, True]), [1, -4])
+        assert_array_equal(result, expected)
+
+        # Condition is single bool tuple
+        expected = numpy.piecewise(a, (True, False), [1])
+        result = dpnp.piecewise(ia, (True, False), [1])
+        assert_array_equal(result, expected)
+
+    def test_error_dpnp(self):
```
Comment (vlad-perevezentsev) [inline dpnp/tests/test_functional.py]: Could you move `test_error_dpnp` and `test_error` to the end of `TestPiecewise`?

File: dpnp/tests/test_functional.py [hunk 15]
```diff
@@ -65,3 +75,244 @@ def custom_func(x, axis):
...
+        # extra function
+        assert_raises_regex(
+            ValueError,
+            "1 or 2 functions are expected",
+            xp.piecewise,
+            ia,
+            [xp.array([True, False])],
+            [1, 2, 3],
+        )
+
```
Comment (vlad-perevezentsev) [inline dpnp/tests/test_functional.py]: Could you add a test with an empty `condlist`?

File: dpnp/tests/test_usm_type.py [hunk 16]
```diff
@@ -755,6 +755,19 @@ def test_apply_over_axes(usm_type):
     assert x.usm_type == y.usm_type
 
 
+@pytest.mark.parametrize("usm_type", list_of_usm_types)
+def test_piecewise(usm_type):
+    x = dpnp.array([0, 0], usm_type=usm_type)
```
Comment (vlad-perevezentsev) [inline dpnp/tests/test_usm_type.py]: Should we test cases where all 3 input arrays have different `usm_type`?

File: pyproject.toml [hunk 17]
```diff
@@ -139,7 +139,7 @@ exclude-protected = ["_create_from_usm_ndarray"]
 
 [tool.pylint.design]
 max-args = 11
-max-branches = 17
+max-branches = 18
```
Comment (vlad-perevezentsev) [inline pyproject.toml:142]: Is this still required?

Comment (vtavana) [reply pyproject.toml:142]: yes

File: dpnp/dpnp_iface_functional.py (hunk 11 above)
Comment (vtavana) [reply dpnp/dpnp_iface_functional.py]: Done!

File: dpnp/dpnp_iface_functional.py (hunk 12 above)
Comment (vtavana) [reply dpnp/dpnp_iface_functional.py]: Done!

File: dpnp/dpnp_iface_functional.py (hunk 13 above)
Comment (vtavana) [reply dpnp/dpnp_iface_functional.py]: Adde a new test

File: dpnp/tests/test_functional.py (hunk 14 above)
Comment (vtavana) [reply dpnp/tests/test_functional.py]: Done!

File: dpnp/tests/test_usm_type.py (hunk 16 above)
Comment (vtavana) [reply dpnp/tests/test_usm_type.py]: Done!

File: dpnp/tests/test_functional.py (hunk 15 above)
Comment (vtavana) [reply dpnp/tests/test_functional.py]: Done!

File: dpnp/dpnp_iface_functional.py (hunk 8 above)
Comment (vtavana) [reply dpnp/dpnp_iface_functional.py:342]: NumPy also raise `IndexError` for this case, so I do not see an immediate benefit to do so.

File: dpnp/dpnp_iface_functional.py (hunk 3 above)
Comment (antonwolfy) [reply dpnp/dpnp_iface_functional.py:342]: No, I mean `condlist` might not be a subscriptable object and so `condlist[0]` would raise `TypeError`

File: dpnp/dpnp_iface_functional.py (hunk 13 above)
Comment (antonwolfy) [inline dpnp/dpnp_iface_functional.py]: need to add exception text

File: dpnp/dpnp_iface_functional.py [hunk 18]
```diff
@@ -266,3 +270,141 @@ def apply_over_axes(func, a, axes):
...
+        dpnp.isscalar(condlist[0]) and x.ndim != 0
+    ):
+        # convert scalar to a list of one array
+        # convert list of scalars to a list of one array
+        condlist = [
+            dpnp.full(
+                x.shape, condlist, usm_type=x.usm_type, sycl_queue=x.sycl_queue
+            )
+        ]
+    elif not isinstance(condlist[0], (dpnp.ndarray)):
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_functional.py]: Should we check for `usm_ndarray` here?

File: dpnp/dpnp_iface_functional.py (hunk 13 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_functional.py]: done

File: dpnp/dpnp_iface_functional.py (hunk 18 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_functional.py]: Yes, I have added it
Thank you!

**Review by antonwolfy — APPROVED** [review]
